aiohttp/web_response.py

```diff
@@ -370,27 +380,36 @@ async def _prepare_headers(self) -> None:
                     "Using chunked encoding is forbidden "
                     "for HTTP/{0.major}.{0.minor}".format(request.version)
                 )
-            writer.enable_chunking()
-            headers[hdrs.TRANSFER_ENCODING] = "chunked"
+            if not self._must_be_empty_body:
+                writer.enable_chunking()
+                headers[hdrs.TRANSFER_ENCODING] = "chunked"
             if hdrs.CONTENT_LENGTH in headers:
                 del headers[hdrs.CONTENT_LENGTH]
         elif self._length_check:
             writer.length = self.content_length
             if writer.length is None:
-                if version >= HttpVersion11 and self.status != 204:
-                    writer.enable_chunking()
-                    headers[hdrs.TRANSFER_ENCODING] = "chunked"
-                    if hdrs.CONTENT_LENGTH in headers:
-                        del headers[hdrs.CONTENT_LENGTH]
-                else:
+                if version >= HttpVersion11:
+                    if not self._must_be_empty_body:
+                        writer.enable_chunking()
+                        headers[hdrs.TRANSFER_ENCODING] = "chunked"
+                elif not self._must_be_empty_body:
                     keep_alive = False
-            # HTTP 1.1: https://tools.ietf.org/html/rfc7230#section-3.3.2
-            # HTTP 1.0: https://tools.ietf.org/html/rfc1945#section-10.4
-            elif version >= HttpVersion11 and self.status in (100, 101, 102, 103, 204):
-                del headers[hdrs.CONTENT_LENGTH]
 
-        if self.status not in (204, 304):
+        # HTTP 1.1: https://tools.ietf.org/html/rfc7230#section-3.3.2
+        # HTTP 1.0: https://tools.ietf.org/html/rfc1945#section-10.4
+        if self._must_be_empty_body:
+            # If the content-length is not set to "0" we will
+            # prematurely close a keep-alive connection that could
+            # have been reused. This matches the behavior of Response
+            # which would also set the content-length to "0" in _start
+            headers[hdrs.CONTENT_LENGTH] = "0"
```

Comment:
If I delete this head instead of setting it to zero the `test_keepalive_after_head_requests_success` test fails.

The test wasn't actually mocking `HEAD` before so it was actually returning an error

Reply:
https://tools.ietf.org/html/rfc7230#section-3.3.2
> A server MAY send a Content-Length header field in a response to a
   HEAD request ([Section 4.3.2 of [RFC7231]](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.2)); a server MUST NOT send
   Content-Length in such a response unless its field-value equals the
   decimal number of octets that would have been sent in the payload
   body of a response if the same request had used the GET method.


So it looks like we have to delete this

Reply:
If you think it improves compatibility with keep-alive clients, it can stay there when it's not a HEAD request. It can't be done for a HEAD request though, as the client may be using the HEAD request to figure out how big the response will be.

Reply:
Ah, actually, 304 is the same. Should be left if present, but shouldn't be set to 0.

It appears that 1xx, 204 are actually forbidden to include the header at all:
https://datatracker.ietf.org/doc/html/rfc9110#section-8.6-8

So, yeah, that will need to go.